File: leetcode/problems/131.ts
function partition(s: string): string[][] {
  const ans: string[][] = [],
    n = s.length
  const check = (i: number, j: number) => {
    const tmp = s.slice(i, j + 1)
    let x = 0,
      y = j - i,
      flag = true
    while (x < y) {
      if (tmp[x] === tmp[y]) {
        ++x, --y
      } else {
        flag = false
        break
      }
    }
    return flag
  }
  const dfs = (idx: number, res: string[]) => {
    if (idx >= n) {
      ans.push(res.slice())
      return
    }
    for (let i = idx; i < n; i++) {
      if (check(idx, i)) {
        res.push(s.slice(idx, i + 1))
        dfs(i + 1, res)
        res.pop()
      }
    }
  }
  dfs(0, [])
  return ans
}

File: leetcode/problems/131.go
func partition(s string) [][]string {
  var ans [][]string
  check := func(i int, j int) bool {
    str := s[i:j+1]
    x, y := 0, len(str)-1
    for x < y {
      if str[x] != str[y] {
        return false
      }
      x ++
      y --
    }
    return true
  }
  var dfs func([]string, int)
  dfs = func(tmp []string, cur int) {
    if cur >= len(s) {
      ans = append(ans, append([]string(nil), tmp...))
      return
    }
    for i := cur; i < len(s); i ++ {
      if check(cur, i) {
        tmp = append(tmp, s[cur:i+1])
        dfs(tmp, i+1)
        tmp = tmp[:len(tmp)-1]
      }
    }
  }
  dfs([]string{}, 0)
  return ans
}